### DISCUSSION TITLE
Expected behaviour of Historical Access for constant data values

### DISCUSSION THREAD

**Discussion by sricharan-code** [FreeOpcUa/opcua-asyncio#613] [Q&A]
Hi everyone,

When I try to perform write operation on a historical variable with same value multiple times, the value gets logged only in the first instance and doesn't log on subsequent attempts. Is this expected behaviour?

**Usecase**: In field, Temperature sensor logs data for every minute. If the temperature remains same for an hour, Is the server supposed to store 60 values or single value?

I tried testing the same with UAExpert. However, the interface doesn't allow me to log the same data.

Please let me know the expected behaviour of sever in such cases.

Thanks in advance.

**Comment by oroulet** [comment]
if you change the sourcetimestamp, then every data_value should be different, but maybe you are writting the same datavalue withtout changing timestamp?

**Reply by sricharan-code** [reply]
I'm using opcua-asyincio client. I used the below code to write the data to historized variable,
`nodeID.write_value(value, ua.VariantType.Int32)`
Correct me if I'm wrong, I guess server will implicitly assign a new timestamp while writing a value using below code in ua_utils.py file.
```
if isinstance(val, ua.DataValue):
        datavalue = val
    elif isinstance(val, ua.Variant):
        datavalue = ua.DataValue(val)
        datavalue.SourceTimestamp = datetime.utcnow()
    else:
        datavalue = ua.DataValue(ua.Variant(val, varianttype))
        datavalue.SourceTimestamp = datetime.utcnow()
    return datavalue
```

**Reply by oroulet** [reply]
yes the client wil create a new timestamp before sending to server.

**Reply by sricharan-code** [reply]
I felt the same and thought that it should work when I write the same data again after some time. Inorder to test this, I used the same write method on loop. However, only the first value has got logged.

**Reply by sricharan-code** [reply]
Hi @oroulet , below code logs the data only if the data change happens.   https://github.com/FreeOpcUa/opcua-asyncio/blob/b482a236de39ff6617920ce444959db1dcceae25/asyncua/server/address_space.py#L716

This contradicts to our earlier assumtion below.

> if you change the sourcetimestamp, then every data_value should be different, but maybe you are writting the same datavalue withtout changing timestamp?

Please advise

**Reply by oroulet** [reply]
well from what we discussed earlier this looks like a bug. but it requires a bit of work to fix correctly. It will change the behavour of subscription and I am not sure what spec says. Should we send a new subscription when a timestamp has changed? If so we can change that. If not we have an issue because hisotry uses subscription and we need a more complicated solution. Can someone check specification for subscriptions?
